[SPIR-V] Add SPV_INTEL_long_composites (#2848)
intel/llvm: sycl/doc/extensions/SPIRV/SPV_INTEL_long_constant_composite.asciidoc

~~~~~~~~~~~~

Modify Section 3.31, "Capability", adding these rows to the Capability table:

--
[options="header"]
|====
2+^| Capability ^| Depends On
| 6089 | *CapabilityLongConstantCompositeINTEL* +
Allow to use *OpTypeStructContinuedINTEL*, *OpConstantCompositeContinuedINTEL*
and *OpSpecConstantCompositeContinuedINTEL* instructions |
|====
--

Instructions
~~~~~~~~~~~~
In section 3.37.6. Type-Declaration Instructions add the new instruction
[cols="3", width="100%"]
|=====
2+^|*OpTypeStructContinuedINTEL* +

Continue specifying an *OpTypeStruct* with number of _Member types_
greater than maximum possible WordCount.

The previous instruction must be an *OpTypeStruct* or an
*OpTypeStructContinuedINTEL* instruction.

_Member types_ follow the same rules as defined for _Member types_ of
*OpTypeStruct*.

| Capability:
*CapabilityLongConstantCompositeINTEL*

| 2 + variable | 6090 | _<id>, <id>, ... Member N type, member N + 1 type_|
|=====

Modify the description of *OpTypeStruct* instruction, adding the
following sentence to the end:
In case if it is not possible to specify all the _member types_ of the structure
by one *OpTypeStruct* instruction, i.e. if number of members of the

> _Result type_ is greater than maximum possible WordCount, the remaining _member

Reviewer: I'm not an expert in articles in the English language(tagging @AGindinson), but
```suggestion
_Result type_ is greater than the maximum possible WordCount, the remaining _member
```

Author: I'm not an expert in them either. Tried to fix.

> types_ are specified by following *OpTypeStructContinuedINTEL* instructions.

Reviewer: ```suggestion
types_ are specified by the following *OpTypeStructContinuedINTEL* instructions.
```

Author: Done.

See <<Specialization, Specialization>>.

| Capability:
*CapabilityLongConstantCompositeINTEL*

| 2 + variable | 6092 | _<id>, <id>, ... Constituents_ |
|=====

Modify the description of *OpConstantComposite* instruction, adding the
following sentence to the end:
In case if it is not possible to specify all the _Constituents_ by one
*OpConstantComposite* instruction, i.e. if number of members of the
_Result type_ and corresponding _Constituents_ is greater than maximum possible
WordCount, the remaining _Constituents_ are specified by following
*OpConstantCompositeContinuedINTEL* instructions.

Modify the description of *OpSpecConstantComposite* instruction, adding the
following sentence to the end:
In case if it is not possible to specify all the _Constituents_ by one
*OpSpecConstantComposite* instruction, i.e. if number of members of the
_Result type_ and corresponding _Constituents_ is greater than maximum possible
WordCount, the remaining _Constituents_ are specified by following
*OpSpecConstantCompositeContinuedINTEL* instructions.

In Section 3.37.8. Memory Instructions, modify the description of the
*OpAccessChain* instruction, adding the following sentence to the end:

> In case _Base_ is an *OpConstantComposite* instruction followed by one or

Reviewer: I think the same is applicable in case _Base_ is an **OpSpecConstantComposite**.

Author: Ok, mentioned that.

more *OpConstantCompositeContinuedINTEL* instructions, the bounds must be bounds

> of *OpConstantComposite* and *OpConstantCompositeContinuedINTEL* instructions

Reviewer: ```suggestion
of *OpConstantComposite* and the following *OpConstantCompositeContinuedINTEL* instructions
```

Author: Done.

> of *OpSpecConstantComposite* and *OpConstantCompositeContinuedINTEL* instructions

Reviewer 2: ```suggestion
of *OpConstantComposite* and *OpConstantCompositeContinuedINTEL* instructions
```

Author: Thanks for the catch!

together.

Issues
------

1) Do we need to define additional validation rules?

Resolution:

No.

2) Do we need modifications of the OpConstantComposite/OpSpecConstantComposite
instruction description?

Resolution:

Yes, it seems that description of these instructions defines one to one match
between composite type members and Constituents by the sentence:
"There must be exactly one Constituent for each top-level
member/element/component/column of the result." Done.

3) We also might want to modify OpAccessChain to clarify how it works on large
constants.

Resolution:

Yes. Added statement clarifying bounds of OpAccessChain instruction in case when
long composite constant is accessed.

Revision History
----------------

[cols="5,15,15,70"]
[grid="rows"]
[options="header"]
|========================================
|Rev|Date|Author|Changes
|1|2020-11-09|Mariya Podchishchaeva|Initial revision

> |2|2020-11-13|Mariya Podchishchaeva|Apply comments from AlexeySo and AlexeySa +

Reviewer: I'd prefer to see my full name as "Alexey Sotkin".

Author: Sorry, fixed.

Contributors
------------

- Mariya Podchishchaeva, Intel
- Alexey Sotkin, Intel
- Ben Ashbaugh, Intel
- Alexey Sachkov, Intel

Notice
------

Copyright (c) 2020 Intel Corporation. All rights reserved.

Status
------

Draft.

Version
-------

[width="40%",cols="25,25"]
|========================================
| Last Modified Date | 2020-12-07
| Revision           | 6
|========================================

Dependencies
------------

This extension is written against the SPIR-V Specification,
Version 1.5 Revision 4, Unified

This extension requires SPIR-V 1.0.

Overview
--------

This extension adds new capability and instructions to allow to represent
composite constants with number of Constituents greater than the maximum
possible WordCount.

Extension Name
--------------

To use this extension within a SPIR-V module, the following
*OpExtension* must be present in the module:

----
OpExtension "SPV_INTEL_long_constant_composite"
----

Modifications to the SPIR-V Specification, Version 1.5
------------------------------------------------------
Capabilities
~~~~~~~~~~~~

Modify Section 3.31, "Capability", adding these rows to the Capability table:

--
[options="header"]
|====

> 2+^| Capability ^| Depends On

Reviewer: In [the latest version of the spec](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_capability_a_capability) the last column titled as "Implicitly Declares"
```suggestion
2+^| Capability ^| Implicitly Declares
```

| 6089 | *CapabilityLongConstantCompositeINTEL* +
Allow to use *OpTypeStructContinuedINTEL*, *OpConstantCompositeContinuedINTEL*
and *OpSpecConstantCompositeContinuedINTEL* instructions |
|====
--

Instructions
~~~~~~~~~~~~
In section 3.37.6. Type-Declaration Instructions add the new instruction
[cols="3", width="100%"]
|=====
2+^|*OpTypeStructContinuedINTEL* +

Continue specifying an *OpTypeStruct* with number of _Member types_
greater than the maximum possible WordCount.

The previous instruction must be an *OpTypeStruct* or an
*OpTypeStructContinuedINTEL* instruction.

_Member types_ follow the same rules as defined for _Member types_ of
*OpTypeStruct*.

| Capability:
*CapabilityLongConstantCompositeINTEL*

| 2 + variable | 6090 | _<id>, <id>, ... Member N type, member N + 1 type_|
|=====

Modify the description of *OpTypeStruct* instruction, adding the
following sentence to the end:
In case if it is not possible to specify all the _member types_ of the structure
by one *OpTypeStruct* instruction, i.e. if number of members of the

> _Result type_ is greater than the maximum possible WordCount, the remaining

Reviewer 3: Please mention, that _Result type_ is a result type of OpConstantComposite instruction, otherwise it looks like that OpTypeStructContinuedINTEL has a result.

Also when the file gets viewed some table rows are missing.

Author: Could you please elaborate what is the problem with table rows? I probably don't see where it is.

Reviewer 3: Nevermind, I missed to where "Modify the description of OpTypeStruct instruction..." belongs.

*OpConstantComposite* instruction, i.e. if number of members of the
_Result type_ and corresponding _Constituents_ is greater than the maximum
possible WordCount, the remaining _Constituents_ are specified by the following
*OpConstantCompositeContinuedINTEL* instructions.

Modify the description of *OpSpecConstantComposite* instruction, adding the
following sentence to the end:
In case if it is not possible to specify all the _Constituents_ by one
*OpSpecConstantComposite* instruction, i.e. if number of members of the
_Result type_ and corresponding _Constituents_ is greater than the maximum
possible WordCount, the remaining _Constituents_ are specified by the following
*OpSpecConstantCompositeContinuedINTEL* instructions.

In Section 3.37.8. Memory Instructions, modify the description of the
*OpAccessChain* instruction, adding the following sentence to the end:
In case _Base_ is an *OpConstantComposite* instruction followed by one or
more *OpConstantCompositeContinuedINTEL* instructions, the bounds must be bounds
of *OpConstantComposite* and the following *OpConstantCompositeContinuedINTEL*
instructions together. Same is applicable in case _Base_ is an

> *OpSpecConstantComposite* instruction followed be one or more

Author: ```suggestion
*OpSpecConstantComposite* instruction followed by one or more
```

possible WordCount, the remaining _Constituents_ are specified by the following
*OpConstantCompositeContinuedINTEL* instructions.

Modify the description of *OpSpecConstantComposite* instruction, adding the
following sentence to the end:
In case if it is not possible to specify all the _Constituents_ by one
*OpSpecConstantComposite* instruction, i.e. if number of members of the
_Result type_ and corresponding _Constituents_ is greater than the maximum
possible WordCount, the remaining _Constituents_ are specified by the following
*OpSpecConstantCompositeContinuedINTEL* instructions.

In Section 3.37.8. Memory Instructions, modify the description of the
*OpAccessChain* instruction, adding the following sentence to the end:
In case _Base_ is an *OpConstantComposite* instruction followed by one or
more *OpConstantCompositeContinuedINTEL* instructions, the bounds must be bounds
of *OpConstantComposite* and the following *OpConstantCompositeContinuedINTEL*
instructions together. Same is applicable in case _Base_ is an
*OpSpecConstantComposite* instruction followed by one or more
*OpSpecConstantCompositeContinuedINTEL* instructions.

> In section 3.37.11. Conversion Instructions, add the new instruction

Reviewer: ```suggestion
In section 3.37.12. Composite Instructions, add the new instruction
```
